### PR TITLE
Core: add snapshot reference to table metadata

### PR DESCRIPTION
**Pull request by amogh-jahagirdar**
Co-authored-by: @hameizi 1249369293@qq.com
Co-authored-by:  @jackye1995 yzhaoqin@amazon.com

This is a continuation of the work already done by Jack and @hameizi in https://github.com/apache/iceberg/pull/3104. Some comments addressed include cleaning up the builder for SnapshotRef, adding the interface methods to get a snapshot ref by name, and validation that only the main branch snapshot reference can have a snapshot id of -1. 

 we will tackle updating snapshot expiration In subsequent PRs as outlined in this doc: https://docs.google.com/document/d/1PvxK_0ebEoX3s7nS6-LOJJZdBYr_olTWH9oepNUfJ-A/edit# .

The functions for handling table mutations and having a reference to the committed mutation will be in a follow on PR as well.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 1]
```diff
@@ -940,9 +1029,7 @@ public Builder setCurrentSnapshot(long snapshotId) {
       return this;
     }
 
-    public Builder removeSnapshots(List<Snapshot> snapshotsToRemove) {
-      Set<Long> idsToRemove = snapshotsToRemove.stream().map(Snapshot::snapshotId).collect(Collectors.toSet());
-
+    public Builder removeSnapshots(Set<Long> idsToRemove) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:1032]: nit: `snapshotIdsToRemove`

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 2]
```diff
@@ -501,9 +521,8 @@ public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
     return new Builder(this).setCurrentSnapshot(snapshot).build();
   }
 
-  public TableMetadata removeSnapshotsIf(Predicate<Snapshot> removeIf) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:504]: I think we still need to keep this method for backwards API compatibility, could you mark it as deprecated instead?

In the original PR I chose to skip removing snapshots with ref/branch, we can discuss what's the preferred behavior if we now decide to deprecate this method.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 3]
```diff
@@ -501,9 +521,8 @@ public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
...
-    List<Snapshot> toRemove = snapshots.stream().filter(removeIf).collect(Collectors.toList());
-    return new Builder(this).removeSnapshots(toRemove).build();
+  public TableMetadata removeSnapshots(Set<Long> snapshotsToRemove) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:524]: I also created https://github.com/apache/iceberg/issues/3900 to track the new API. Not sure what is the preference, but I think it's better to have a PR that is dedicated to the new APIs and old method deprecation, because this PR is already quite long.

File: api/src/main/java/org/apache/iceberg/SnapshotRef.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+  /**
+   * Creates a ref builder from the given ref and its properties but the ref will now point to the given snapshotId.
+   * @param ref Ref to build from
+   * @param snapshotId snapshotID to use.
+   * @return ref builder with the same retention properties as given ref, but the ref will point to the passed in id
+   */
+  public static Builder builderFrom(SnapshotRef ref, long snapshotId) {
+    return new Builder(ref.type(), snapshotId)
+            .minSnapshotsToKeep(ref.minSnapshotsToKeep())
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/SnapshotRef.java:94]: Can you fix indentation for these lines?

File: api/src/main/java/org/apache/iceberg/SnapshotRef.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            "Tags do not support setting maxSnapshotAgeMs");
+      } else {
+        ValidationException.check(minSnapshotsToKeep == null || minSnapshotsToKeep > 0,
+                "Min snapshots to keep must be greater than 0");
+        ValidationException.check(maxSnapshotAgeMs == null || maxSnapshotAgeMs > 0,
+                "Max snapshot age must be greater than 0");
+      }
+
+      if (maxRefAgeMs != null) {
+        ValidationException.check(maxRefAgeMs > 0, "Max reference age must be greater than 0");
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/SnapshotRef.java:148]: This can be rewritten to avoid the unnecessary `if`:

```java
ValidationException.check(maxRefAgeMs == null || maxRefAgeMs > 0, "Max reference age must be greater than 0");
```

File: api/src/main/java/org/apache/iceberg/SnapshotRef.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+      this.maxRefAgeMs = value;
+      return this;
+    }
+
+    public SnapshotRef build() {
+      if (type.equals(SnapshotRefType.TAG)) {
+        ValidationException.check(minSnapshotsToKeep == null,
+            "Tags do not support setting minSnapshotsToKeep");
+        ValidationException.check(maxSnapshotAgeMs == null,
+            "Tags do not support setting maxSnapshotAgeMs");
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/SnapshotRef.java:139]: Why not enforce these when calling `maxSnapshotAgeMs` or `minSnapshotsToKeep`? If builders are passed around, then those are probably better places to throw exceptions.

I'd probably move these checks to those methods are `Precondition` checks.

**Comment by amogh-jahagirdar** [comment]
Thanks for taking a look everyone, appreciate it! @rdblue @jackye1995 @RussellSpitzer please let me know if you have any other feedback.

File: api/src/main/java/org/apache/iceberg/SnapshotRef.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+  public static Builder builderFrom(SnapshotRef ref) {
+    return new Builder(ref.type(), ref.snapshotId())
+        .minSnapshotsToKeep(ref.minSnapshotsToKeep())
+        .maxSnapshotAgeMs(ref.maxSnapshotAgeMs())
+        .maxRefAgeMs(ref.maxRefAgeMs());
+  }
+  /**
+   * Creates a ref builder from the given ref and its properties but the ref will now point to the given snapshotId.
+   * @param ref Ref to build from
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/SnapshotRef.java:86]: nit: newline after javadoc description

File: api/src/main/java/org/apache/iceberg/Table.java [hunk 8]
```diff
@@ -138,6 +138,21 @@ default String name() {
...
+   */
+  Map<String, SnapshotRef> refs();
+
+  /**
+   * Get the snapshot reference with the given name
+   *
+   * @param refName snapshot reference name
+   * @return the SnapShot ref with the given name if it exists, null otherwise.
+   */
+  SnapshotRef ref(String refName);
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/Table.java:154]: nit: prefer simple variable name, `name` instead of `refName` since `ref` is implied from the method name.

File: api/src/main/java/org/apache/iceberg/Table.java [hunk 9]
```diff
@@ -138,6 +138,21 @@ default String name() {
...
+   *
+   * @return a map with ref name as key, {@link SnapshotRef} as value
+   */
+  Map<String, SnapshotRef> refs();
+
+  /**
+   * Get the snapshot reference with the given name
+   *
+   * @param refName snapshot reference name
+   * @return the SnapShot ref with the given name if it exists, null otherwise.
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/Table.java:152]: nit: lower case snapshot

File: api/src/main/java/org/apache/iceberg/Table.java [hunk 10]
```diff
@@ -138,6 +138,21 @@ default String name() {
    */
   List<HistoryEntry> history();
 
+  /**
+   * Get the snapshot references of this table.
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/Table.java:142]: worth pointing out that there is always a snapshot ref for the main branch. If there is no snapshot in the table, main branch points to -1

File: api/src/main/java/org/apache/iceberg/UpdateSnapshotRefs.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+package org.apache.iceberg;
+
+import java.util.Map;
+import org.apache.iceberg.exceptions.CommitFailedException;
+
+/**
+ * API for snapshot reference evolution.
+ * <p>
+ * When committing, these changes will be applied to the current table metadata. Commit conflicts
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/UpdateSnapshotRefs.java:28]: nit: prefer full sentence in javadoc, change to a newline after `.`

File: core/src/main/java/org/apache/iceberg/BaseTable.java [hunk 12]
```diff
@@ -124,6 +124,16 @@ public Snapshot snapshot(long snapshotId) {
     return ops.current().snapshotLog();
   }
 
+  @Override
+  public Map<String, SnapshotRef> refs() {
+    return ops.current().refs();
+  }
+
+  @Override
+  public SnapshotRef ref(String refName) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/BaseTable.java:133]: nit: `name` as suggested before

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 13]
```diff
@@ -587,6 +595,16 @@ public Snapshot snapshot(long snapshotId) {
       return current.snapshotLog();
     }
 
+    @Override
+    public Map<String, SnapshotRef> refs() {
+      return current.refs();
+    }
+
+    @Override
+    public SnapshotRef ref(String refName) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:604]: nit: `name` as suggested before

File: core/src/main/java/org/apache/iceberg/SerializableTable.java [hunk 14]
```diff
@@ -262,6 +262,16 @@ public Snapshot snapshot(long snapshotId) {
     return lazyTable().history();
   }
 
+  @Override
+  public Map<String, SnapshotRef> refs() {
+    return lazyTable().refs();
+  }
+
+  @Override
+  public SnapshotRef ref(String refName) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SerializableTable.java:271]: nit: `name` as suggested before

File: core/src/main/java/org/apache/iceberg/SnapshotReferenceParser.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to parse snapshot ref: " + json, e);
+    }
+  }
+
+  public static SnapshotRef fromJson(JsonNode node) {
+    Preconditions.checkArgument(node.isObject(), "Cannot parse snapshot reference from a non-object: %s", node);
+    long snapshotId = JsonUtil.getLong(SNAPSHOT_ID, node);
+    SnapshotRefType type = SnapshotRefType.valueOf(
+        JsonUtil.getString(TYPE, node).toUpperCase(Locale.ENGLISH));
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotReferenceParser.java:84]: does this need to be on the newline?

File: core/src/test/java/org/apache/iceberg/TestUpdateSnapshotRefs.java [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+        () -> table.updateRefs().remove(SnapshotRef.MAIN_BRANCH));
+  }
+
+  @Test
+  public void testRemoveRefNotExist() {
+    AssertHelpers.assertThrows("Should fail when removing ref not exist",
+        IllegalArgumentException.class,
+        "Cannot find ref to remove",
+        () -> table.updateRefs().remove("tag"));
+  }
```
Comment (jackye1995) [inline core/src/test/java/org/apache/iceberg/TestUpdateSnapshotRefs.java:254]: I think we are missing a test for chaining multiple operations for tag/branch, like

```
table.updateRefs()
  .branch("b", 123)
  .setMinSnapshotsInBranch("b", 10000)
  .build()
```

to make sure chaining works as expected when creating new ref and set values at the same time.

**Review by jackye1995 — COMMENTED** [review]
mostly looks good to me, just nitpicking

File: core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+    this.ops = ops;
+    this.base = ops.current();
+    this.refs = Maps.newHashMap(base.refs());
+  }
+
+  @Override
+  public UpdateSnapshotRefs tag(String name, long snapshotId) {
+    Preconditions.checkArgument(name != null, "Tag name must not be null");
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null, "Cannot find snapshot with ID: %s", snapshotId);
+    Preconditions.checkArgument(!refs.containsKey(name), "Cannot tag snapshot, ref already exists: %s", name);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:42]: Should there be a `replaceTag` method for this case? Otherwise, I'd have to remove a tag and re-add it, but I wouldn't really know whether the tag already existed or if it was a tag and not a branch.

File: core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+    refs.put(name, SnapshotRef.tagBuilder(snapshotId).build());
+    return this;
+  }
+
+  @Override
+  public UpdateSnapshotRefs branch(String name, long snapshotId) {
+    Preconditions.checkArgument(name != null, "Branch name must not be null");
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null, "Cannot find snapshot with ID: %s", snapshotId);
+    Preconditions.checkArgument(!refs.containsKey(name), "Cannot create branch, ref already exists: %s", name);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:52]: Looks like this doesn't support updating a branch? Is the intent that committing to a branch is an operation handled elsewhere instead?

File: core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java (hunk 17 above)
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:42]: yeah the current API tries to have just `create`, `remove` and `rename` and ref property update APIs, because it's a bit unclear to me if we need a replace given that there is no such operation in schema and partition spec update.

The action `replace` indicates the tag already exists, and I imagine it should throw exception if the tag does not exist. In that case, user still need to first check existence and then decide to create or replace, and it has little difference from just updating properties of the ref.

If we allow `replace` to mean `put` that either create or update, then my concern is that it might cause unintented use that people might override a tag without knowing it already exists. The current API is basically trying to force a check on user side because there is not really a compare-and-swap of the specific ref that we could enforce as of today on server side and we have to replace the entire metadata file.

any thoughts?

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:42]: I need to think more about what we want in this API vs in the other APIs. Probably worth thinking through the TableMetadata changes and API as well. It could be that we want to have this style API there instead of here and just keep this to high level operations.

File: core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java (hunk 18 above)
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:52]: oh yes good point. Currently the expectation is that `TableMetadata` always updates the `main` branch to be the current snapshot ID so that operation is not needed. But it will cumbersome for a user to do remove and recreate the branch just to update the branch head. Do you think this is something we can add in another PR together with updates to the snapshot producers, or better to be out in the same one?

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:52]: I'm thinking about what the API for the TableMetadata builder should be. Then we can kind of work backwards from there. I'm wondering how we set the operating branch when committing, too. Right now we have a `stageOnly` method on `SnapshotProducer`. We could do something similar. Maybe update that to produce tagged snapshots?

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:52]: From the snapshot producer perspective, I was thinking 2 cases:

1) Append to branch:
```
table.newRewrite()
            .rewriteFiles(Sets.newSet(FILE_A), Sets.newSet(FILE_B))
            .atBranch("beta")
```

In this case, I would say both use cases make sense:
1. user might want to fail when committing to a non-existing branch, because that branch was deleted by someone else. This could be satisfied by the current API by removing old branch and update it with the same info, a bit cumbersome but still works.
2. user might want to create the branch with the new snapshot ID, this could also be done by the current API by first checking branch existence and then do a create after snapshot ID is committed.

2) Add tag:

```
table.newRewrite()
            .rewriteFiles(Sets.newSet(FILE_A), Sets.newSet(FILE_B))
            .tag("2022-01-01-compacted"）
```

In this case, tagging with an existing name is likely a system issue as there should only be a single process doing the commit and tag it. This could be identified by the current API.

So from my perspective the current API is enough, although the branch interface is not straight forward for people to just update the branch head. I would imagine it to be a frequent operation, so might worth adding that.

Regarding the builder API, currently it's following the pattern of property update. I think `set` and `remove` would be beneficial for performing conditional update for specific refs and avoid full metadata check as we develop the rest catalog, do you think there is any other APIs for the builder that would be more suitable?

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java:52]: Yes, I think you're right. There are use cases for both branching and tagging in the snapshot producer updates. Here's what I'm thinking for the API:

* `SnapshotProducer.stageOnly()` - like the current behavior, creates a snapshot without updating a branch
* `SnapshotProducer.forBranch(String name)` - commit changes to the given branch
  * Defaults to "main" branch if not called
  * Incompatible with `stageOnly()`
  * If the branch does not exist, it will be created from main
* `SnapshotProducer.withTag(String name)` - tag the committed snapshot as name
  * May be used with `stageOnly()` or `forBranch(...)`

This behavior differs from what you were talking about because this will create the branch if it doesn't exist, which means that you wouldn't throw an exception if someone dropped a branch that a job was writing to. I think that's okay because writes are expensive. The best thing to do is to accept the new snapshot and correctly label it -- the user can always drop it easily if it is actually unwanted. Another reason to do this is to make it easy to create branches. You just write to a branch and it forks from master.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 19]
```diff
@@ -342,6 +349,8 @@ public String toString() {
     Preconditions.checkArgument(
         currentSnapshotId < 0 || snapshotsById.containsKey(currentSnapshotId),
         "Invalid table metadata: Cannot find current version");
+
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:353]: Unnecessary whitespace change.

File: core/src/main/java/org/apache/iceberg/MetadataUpdate.java [hunk 20]
```diff
@@ -200,4 +200,28 @@ public String location() {
       return location;
     }
   }
+
+  class SetSnapshotRefs implements MetadataUpdate {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MetadataUpdate.java:204]: I think this API should focus on updates to individual refs, rather than updating in bulk. What I added to the OpenAPI REST spec was `SetSnapshotRefUpdate`, which is a JSON object like this:

```json
{
  "action": "set-snapshot-ref",
  "ref": "ref name",
  "type": "tag|branch",
  "snapshot-id": 50120357190375
}
```

These can optionally contain `max-ref-age-ms`, `max-snapshot-age-ms`, and `min-snapshots-to-keep`.

File: core/src/main/java/org/apache/iceberg/MetadataUpdate.java [hunk 21]
```diff
@@ -200,4 +200,28 @@ public String location() {
...
+    public SetSnapshotRefs(Map<String, SnapshotRef> updated) {
+      this.updated = updated;
+    }
+
+    public Map<String, SnapshotRef> updated() {
+      return updated;
+    }
+  }
+
+  class RemoveSnapshotRefs implements MetadataUpdate {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MetadataUpdate.java:216]: Like `RemoveSnapshot` and `SetSnapshotRef`, I would probably make this remove a single ref.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 22]
```diff
@@ -472,6 +481,18 @@ public Snapshot currentSnapshot() {
     return previousFiles;
   }
 
+  public Map<String, SnapshotRef> refs() {
+    return refs;
+  }
+
+  public Set<SnapshotRef> refs(long snapshotId) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:488]: Is this used anywhere?

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 23]
```diff
@@ -502,10 +523,15 @@ public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
   }
 
   public TableMetadata removeSnapshotsIf(Predicate<Snapshot> removeIf) {
-    List<Snapshot> toRemove = snapshots.stream().filter(removeIf).collect(Collectors.toList());
+    Set<Long> toRemove = snapshots.stream()
+            .filter(removeIf)
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:527]: Please fix indentation. Continuing indentation is 4 spaces (2 indents).

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 24]
```diff
@@ -502,10 +523,15 @@ public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
...
+            .filter(snapshot -> !refsById.containsKey(snapshot.snapshotId()))
+            .map(Snapshot::snapshotId)
+            .collect(Collectors.toSet());
     return new Builder(this).removeSnapshots(toRemove).build();
   }
 
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:534]: Unnecessary whitespace change.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 25]
```diff
@@ -529,6 +555,24 @@ public TableMetadata replaceProperties(Map<String, String> rawProperties) {
         .build();
   }
 
+  public TableMetadata replaceRefs(Map<String, SnapshotRef> newRefs) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:558]: I don't think that replacing all refs is a good idea. In fact, I would probably avoid adding new mutation methods to `TableMetadata` in general and have you go directly to the `TableMetadata.Builder` API.

File: core/src/main/java/org/apache/iceberg/TableMetadataParser.java [hunk 26]
```diff
@@ -238,6 +241,13 @@ private static void toJson(TableMetadata metadata, JsonGenerator generator) thro
     }
     generator.writeEndArray();
 
+    generator.writeObjectFieldStart(REFS);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadataParser.java:244]: I think this should be written before the list of snapshots.

File: core/src/main/java/org/apache/iceberg/util/JsonUtil.java [hunk 27]
```diff
@@ -134,6 +147,18 @@ public static String getStringOrNull(String property, JsonNode node) {
         .build();
   }
 
+  public static void writeIntegerIfExists(String key, Integer value, JsonGenerator generator) throws IOException {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/JsonUtil.java:150]: In other places with a similar pattern, we move the whole boolean to the caller so that this is a bit more flexible:

```java
  JsonUtil.writeIntegerIf(v != null, "min-snapshots-to-keep", v, generator);
```

**Comment by rdblue** [comment]
@amogh-jahagirdar, there are a lot of good changes in here, but I think we should break this up into smaller, more manageable commits:

* Parser updates to read/write refs
* Additions to the Table API
* The ref management API
* TableMetadata updates
* Updates to SnapshotProducer API
* Updates to MergingSnapshotProducer and FastAppend

Can you refactor the parser updates and API interface updates into separate PRs? In the meantime, I'll take a look at the table metadata updates.

**Comment by amogh-jahagirdar** [comment]
@rdblue Sure, makes sense! I will break this up into separate PRs.

**Comment by rdblue** [comment]
I think most of this is in, so I'm going to close this one. Thanks, @amogh-jahagirdar!
